DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.hpp

```diff
@@ -0,0 +1,1852 @@
+#include <dirent.h>
+#include <limits.h>
+#include <CL/sycl.hpp>
+#include <vector>
+
+#include "host_speed.hpp"
+
+// Pre-declare kernel name to prevent name mangling
+// This is an FPGA best practice that makes it easier to identify the kernel in
+// the optimization reports.
+class NopNDRange;
+class NopSingleTask;
+class KernelSender;
+class KernelReceiver;
+class MemReadWriteStream;
+class MemReadWriteStreamNDRange;
+class MemWriteStream;
+class MemReadStream;
+
+// Pipe used between KernelSender and KernelReceiver -
+// ShimMetrics::KernelLaunchTest(queue &q) function
+using SendertoReceiverPipe = INTEL::pipe<  // Defined in the SYCL headers
+    class SenderReceiverPipe,              // An identifier for the pipe
+    unsigned int,                          // The type of data in the pipe
+    1>;                                    // The capacity of the pipe
+
+/////////////////////////////////
+// **** class ShimMetrics **** //
+/////////////////////////////////
+
+// Object stores oneAPI shim metrics
+// Member Functions (details closer to function definition):
+// ShimMetrics - Constructor; initializes all metrics and obtains maximum device
+// allocation and maxmum device global memory TestGlobalMem - Host to device
+// global memory interface check HostSpeed - Host to device global memory
+// bandwidth measurement HostRWTest - Unaligned read & writes from host to
+// device global memory KernelClkFreq - Kernel clock frequency measurement
+// KernelLatency - Kernel latency measurement
+// KernelLaunchTest - Host to kernel interface check
+// KernelMemRW - Kernel to device global memory interface check
+// KernelMemBW - Kernel to device global memory bandwidth measurement
+
+class ShimMetrics {
+ public:
+  ShimMetrics(queue &q)
+      : h2d_rd_bw{0},
+        h2d_wr_bw{0},
+        h2d_rd_wr_bw{0},
+        h2d_rw_test{false},
+        kernel_freq{0},
+        kernel_latency{0},
+        kernel_thruput{0},
+        kernel_mem_bw{0},
+        kernel_mem_rw_test{false} {
+    max_buffer_size = q.get_device().get_info<info::device::global_mem_size>();
+    max_alloc_size =
+        q.get_device().get_info<info::device::max_mem_alloc_size>();
+    std::cout << "\nclGetDeviceInfo CL_DEVICE_GLOBAL_MEM_SIZE = "
+              << max_buffer_size << "\n";
+    std::cout << "clGetDeviceInfo CL_DEVICE_MAX_MEM_ALLOC_SIZE = "
+              << max_alloc_size << "\n";
+    std::cout << "Device buffer size available for allocation = "
+              << max_alloc_size << " bytes\n";
+  }
+
+  ~ShimMetrics() {}
+
+  size_t TestGlobalMem(queue &q);
+  int HostSpeed(queue &q);
```

Comment:
why do these functions require a reference to `q`? since that reference was passed to the `ShimMetrics` constructor, can the reference to `q` be a member like `kernel_mem_bw`?

Reply:
You could store a reference in this class, but it may be destructed behind your back. I am not convinced that these set of tests should be a class. If we kept it a class, I agree with Paul that `q` should be a member.